### DISCUSSION TITLE
Passing parameters to HYPRE

### DISCUSSION THREAD

**Discussion by konstpas** [AMReX-Codes/amrex#2924] [Q&A]
Hi,

I'm trying to use a hypre solver as the bottom solver in my fortran code by calling multigrid%set_bottom_solver(3). I think the hypre solver is indeed called because when I use the default AMReX solver the linear solver diverges while when I set it to hypre it converges. I do not see the output shown in the example ABecLaplacian_C (HYPRE BoomerAMG: Num. iterations = ..; Relative residual = .....) but I assume this is due to the verbosity of the bottom solver.

My issue is that I cannot pass parameters to hypre by following what is said on the [source documentation page](https://amrex-codes.github.io/amrex/docs_html/LinearSolvers.html#external-solvers) i.e. setting amrex.init_hypre=1 in the input followed by arguments of the hypre.bamg_verbose=1

In case it is relevant, I built AMReX as an external library and tried different configure options:
1) --with-mpi=no --with-omp=yes
2) --with-mpi=no --with-omp=yes --enable-hypre=yes
3) --with-mpi=yes --with-omp=yes --enable-hypre=yes
In cases 1 and 3 my code can compile and run. My code is not written to be parallelized with MPI so I would prefer to build AMReX without mpi. In case 2 I get errors during (I think) the linking of unidentified references, e.g. "/home/konstantinos/amrex/Src/Extern/HYPRE/AMReX_HypreNodeLap.cpp:308: undefined reference to `HYPRE_IJVectorSetValues'"

Thank you in advance!

**Comment by WeiqunZhang** [comment]
Could you try `hypre.verbose=1`? `hypre.bamg_verbose` is for when BoomerAMG is used as a preconditioner.  You can also call `multigrid % set_bottom_verbose(2)` to ask AMReX to print some information regarding the bottom solver.

As for the undefined reference when MPI is not used, I don't know.  That's a hypre function.  Maybe it has something to do with how hypre is compiled.  Nevertheless, you can always run on just one process and your serial should be just fine even if it's compiled with MPI.

**Reply by konstpas** [reply]
Unfortunately, If I set `hypre.verbose=1` then I see that at the output `hypre.verbose` has been added to the list of unused ParmParse variables along with `amrex.init_hypre`.

 Trying `multigrid % set_bottom_verbose(2)` did not produce any additional output  while the bottom solver was hypre. Of course I see extra output if I set `multigrid%set_bottom_solver(1)`.

**Reply by WeiqunZhang** [reply]
If `amrex.init_hypre` is an unused parameter, I think that means amrex is not complied with hypre support.

**Reply by WeiqunZhang** [reply]
Anyway, this is what I did.
```
amrex/Tests/LinearSolvers/ABecLaplacian_F$ make -j6 USE_HYPRE=TRUE
./main3d.gnu.MPI.ex inputs-rt-abeclap-lev bottom_solver=3 hypre.verbose=1 bottom_verbose=2
```
Then I see things like
```
 Num MPI tasks = 1

 Num OpenMP threads = 1


BoomerAMG SETUP PARAMETERS:

 Max levels = 25
 Num levels = 2

 Strength Threshold = 0.570000
 Interpolation Truncation Factor = 0.000000
 Maximum Row Sum Threshold for Dependency Weakening = 0.900000

 Coarsening Type = Falgout-CLJP 
 measures are determined locally


 No global partition option chosen.

 Interpolation = modified classical interpolation

Operator Matrix Information:

             nonzero            entries/row          row sums
lev    rows  entries sparse   min  max     avg      min         max
======================================================================
  0       8       32  0.500     4    4     4.0   1.036e-05   1.036e-05
  1       4       16  1.000     4    4     4.0   2.072e-05   2.072e-05


Interpolation Matrix Information:
                    entries/row        min        max            row sums
lev  rows x cols  min  max  avgW     weight      weight       min         max
================================================================================
  0     8 x 4       1    3   3.0   3.333e-01   3.333e-01   1.000e+00   1.000e+00


     Complexity:    grid = 1.500000
                operator = 1.500000
                memory = 2.000000




BoomerAMG SOLVER PARAMETERS:

  Maximum number of cycles:         200 
  Stopping Tolerance:               1.000000e-04 
  Cycle type (1 = V, 2 = W, etc.):  1

  Relaxation Parameters:
   Visiting Grid:                     down   up  coarse
            Number of sweeps:            2    2     1 
   Type 0=Jac, 3=hGS, 6=hSGS, 9=GE:      6    6     9 
   Point types, partial sweeps (1=C, -1=F):
                  Pre-CG relaxation (down):   1  -1   1  -1
                   Post-CG relaxation (up):  -1   1  -1   1
                             Coarsest grid:   0

HYPRE BoomerAMG: Num. iterations = 1; Relative residual = 3.169823357e-15
MLMG: Iteration   1 Fine resid/bnorm = 1.024235372
```
